File: mcp-server/src/handlers/cursor.ts
import { callNeovim } from '../rpc.js';

/**
 * Retrieve the current Neovim cursor position and return it as pretty-printed JSON in a content block.
 *
 * @returns An object with a `content` array containing a single text node whose `text` is the pretty-printed JSON of the cursor position.
 */
export async function handleGetCursor(args: any) {
  const pos = await callNeovim('get_cursor_position', {}, args?.rpc_port);
  return {
    content: [{ type: 'text', text: JSON.stringify(pos, null, 2) }],
  };
}

/**
 * Move Neovim's cursor to the specified line and optional column.
 *
 * @param args - Object containing cursor position; must include `line` (number). `col` (number) and
 * `winnr` (number) are optional. Without `winnr` the currently active window is moved, which is
 * not the one `nvim_win_open_file` just opened a file in — it restores focus before returning.
 * @returns An object with a `content` array containing a single text node confirming the new cursor line.
 * @throws Error if `args` is falsy or `args.line` is undefined with message 'Missing required parameter: line'.
 */
export async function handleSetCursor(args: any) {
  if (!args || args.line === undefined) {
    throw new Error('Missing required parameter: line');
  }
  await callNeovim(
    'set_cursor_position',
    {
      line: args.line,
      col: args.col,
      winnr: args.winnr,
    },
    args?.rpc_port
  );
  const where = args.winnr === undefined ? '' : ` in window ${args.winnr}`;
  return {
    content: [{ type: 'text', text: `Cursor moved to line ${args.line}${where}` }],
  };
}

/**
 * Fetches the current visual selection from Neovim and wraps its pretty-printed JSON in a content text node.
 *
 * @returns An object with a `content` array containing a single text node whose `text` is the pretty-printed JSON representation of the visual selection
 */
export async function handleGetVisualSelection(args: any) {
  const selection = await callNeovim('get_visual_selection', {}, args?.rpc_port);
  return {
    content: [{ type: 'text', text: JSON.stringify(selection, null, 2) }],
  };
}

File: lua/vibing/infrastructure/rpc/handlers/cursor.lua
local M = {}

-- Get the current window cursor position as a table with fields `line` and `col`.
-- @return table with fields:
--   - `line`: 1-based line number.
--   - `col`: 0-based column index.
function M.get_cursor_position(params)
  local pos = vim.api.nvim_win_get_cursor(0)
  return {
    line = pos[1],
    col = pos[2],
  }
end

-- Set a window's cursor to the specified position.
--
-- `winnr` matters more than it looks. Window 0 is the *current* window, which is not the one
-- `win_open_file` just opened a file in — that handler restores focus before returning. So the
-- documented open-then-jump sequence ("show me the code", every stop of the code-tour skill)
-- moved the chat's cursor instead of the opened file's until this took a window.
--
-- @param params Table with cursor position fields:
--   - line (number): 1-based line number to move the cursor to (required).
--   - col (number): 0-based column number within the line (optional, defaults to 0).
--   - winnr (number): window to move (optional, defaults to the current window).
-- @return table A table `{ success = true }` on successful cursor move.
-- @throws error if `params.line` is not provided, or `params.winnr` is not a valid window.
function M.set_cursor_position(params)
  local line = params and params.line
  local col = params and params.col or 0
  local winnr = params and params.winnr
  if not line then
    error("Missing line parameter")
  end
  if winnr and not vim.api.nvim_win_is_valid(winnr) then
    error("Invalid window number: " .. tostring(winnr))
  end
  vim.api.nvim_win_set_cursor(winnr or 0, { line, col })
  return { success = true }
end

-- Retrieves the current visual selection's lines and positions.
-- @return table containing:
--   lines: array of strings for each selected line
--   start: position list for the selection start (as returned by `vim.fn.getpos`)
--   ["end"]: position list for the selection end (as returned by `vim.fn.getpos`)
function M.get_visual_selection(params)
  local start_pos = vim.fn.getpos("'<")
  local end_pos = vim.fn.getpos("'>")
  local lines = vim.fn.getline(start_pos[2], end_pos[2])
  return {
    lines = lines,
    start = start_pos,
    ["end"] = end_pos,
  }
end

return M
